### ISSUE TITLE
mofchecker instatiation already requires creating graph

### ISSUE THREAD

**Issue by ltalirz** [kjappelbaum/mofchecker#144]
Instantiating the `MOFChecker` class already requires the graph to be constructed, e.g. because it leads to a call `BaseMissingCheck.from_mofchecker`

https://github.com/kjappelbaum/mofchecker/blob/a0cb3051785bf8bbdca811c95794b49a5b4f65e5/mofchecker/checks/local_structure/base_missing_check.py#L25-L31

That does not seem like a good idea and may be part of the reason for https://github.com/kjappelbaum/mofchecker/issues/133

@kjappelbaum can you propose a fix for this?

**Comment by kjappelbaum** [comment]
so the use case would be that you want to have a charge/porosity/has_metal ... check and this should then not compute the graph? 

the easy fix, which I wouldn't like to go for, is that the user specifies which checks are needed and then we might not build the graph.

the other solution would be that we delay the initialization of the objects until they are called and construct the graph upon the first call.

**Comment by ltalirz** [comment]
> so the use case would be that you want to have a charge/porosity/has_metal ... check and this should then not compute the graph?

Correct.

> the other solution would be that we delay the initialization of the objects until they are called and construct the graph upon the first call.

Right, don't compute the graph until it's needed

**Comment by ltalirz** [comment]
Also, it is new to me that the creating of the networkx graph is so expensive, was that the case already before?

The expensive thing is the call to `nx.minimum_cycle_basis(graph)`.
According to [this](https://link.springer.com/article/10.1007/s00453-007-9064-z), for a "reasonably dense" graph with m edges and n vertices and unweighted edges (am I right in assuming that this is the case here?) the cost should be of O(m^ω) with ω<2.376.

I'm seeing a somewhat worse scaling in quick experiments (but perhaps it's still within reason):

* str_m1_o12435_NPO.cif: 174 atoms, 231 edges: 4.5 seconds spent
* str_m1_o12435_ASV.cif: 290 atoms, 378 edges: 24 seconds spent in `minimum_cycle_basis`
  Expecting (378/231)**2.376 * 4.5 = 14.5s
* str_m1_o12435_ATS.cif: 348 atoms, 457 edges: 41.9s  spent in `minimum_cycle_basis`
  Expecting (457/378)**2.376 * 24 = 37.7s

However, this means of course that the MOF checker can never be fast
E.g. for a MOF with 2000 atoms, with 2400 edges: (2400/457)**2.376*42=36 minutes.

**Comment by kjappelbaum** [comment]
yes, minimum basis is expensive (but still much better than a full graph isomorphism, which is NP)

**Comment by ltalirz** [comment]
So, with the latest master the networkx cycle computation is no longer performed upon initialization (which is great), but the BaseMissingCheck still computes the (pymatgen) structure graph.

I guess we would like to avoid that?
As mentioned, I think it would make sense to add to some unit test of "cheap" descriptors an `assert self._graph is None`
